### PR TITLE
Added new event LoginFailEvent

### PR DESCRIPTION
**Pull request by Nan1t**
Added new event to listen when the login process has failed.

### CODE REVIEW

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/LoginSessionHandler.java [hunk 1]
```diff
@@ -127,16 +128,25 @@ public boolean handle(EncryptionResponse packet) {
           } else if (profileResponse.getStatusCode() == 204) {
             // Apparently an offline-mode user logged onto this online-mode proxy.
             inbound.disconnect(VelocityMessages.ONLINE_MODE_ONLY);
+            server.getEventManager().fireAndForget(new LoginFailEvent(inbound,
```
Comment (astei) [inline proxy/src/main/java/com/velocitypowered/proxy/connection/client/LoginSessionHandler.java:131]: Consider a `disconnect()` private function for cases like these

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/LoginSessionHandler.java [hunk 2]
```diff
@@ -225,6 +237,12 @@ private void initializePlayer(GameProfile profile, boolean onlineMode) {
   }
 
   private void finishLogin(ConnectedPlayer player) {
+    Optional<RegisteredServer> toTry = player.getNextServerToTry();
+    if (!toTry.isPresent()) {
+      player.disconnect(VelocityMessages.NO_AVAILABLE_SERVERS);
```
Comment (astei) [inline proxy/src/main/java/com/velocitypowered/proxy/connection/client/LoginSessionHandler.java:242]: You forgot this case

File: api/src/main/java/com/velocitypowered/api/event/connection/LoginFailEvent.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+    public Reason getReason() {
+        return reason;
+    }
+
+    public enum Reason {
+
+        ONLINE_MODE_ONLY,
+        ALREADY_CONNECTED,
+        MANUALLY,
```
Comment (astei) [inline api/src/main/java/com/velocitypowered/api/event/connection/LoginFailEvent.java:36]: This is not a very good name - a more accurate one would be `PLUGIN`

File: api/src/main/java/com/velocitypowered/api/event/connection/LoginFailEvent.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        UNDEFINED,
+
```
Comment (astei) [inline api/src/main/java/com/velocitypowered/api/event/connection/LoginFailEvent.java:38]: Consider adding a new case for a player-initiated disconnect

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/LoginSessionHandler.java (hunk 2 above)
Comment (Leymooo) [reply proxy/src/main/java/com/velocitypowered/proxy/connection/client/LoginSessionHandler.java:242]: i think this should not be in this pr. https://github.com/VelocityPowered/Velocity/commit/d2b8271eb4b23578d833058af4e2fb1cbff4ba94#diff-8e51e6774e951b58c22066197a838ad4

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/LoginSessionHandler.java (hunk 1 above)
Comment (Naanit) [reply proxy/src/main/java/com/velocitypowered/proxy/connection/client/LoginSessionHandler.java]: Done in the new commit. Thanks for the tip.

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/LoginSessionHandler.java (hunk 2 above)
Comment (astei) [reply proxy/src/main/java/com/velocitypowered/proxy/connection/client/LoginSessionHandler.java:236]: Agreed. This reverts a feature added in 1.1.0 to allow users to configure the proxy to run without servers configured. Please revert this.

File: proxy/src/main/java/com/velocitypowered/proxy/connection/client/LoginSessionHandler.java [hunk 5]
```diff
@@ -289,4 +305,8 @@ public void disconnected() {
       connectedPlayer.teardown();
     }
   }
+
+  private void disconnect(String username, LoginFailEvent.Reason reason){
```
Comment (astei) [inline proxy/src/main/java/com/velocitypowered/proxy/connection/client/LoginSessionHandler.java:309]: When I suggested this function name, I also wanted it to take over the responsibility of also disconnecting the user, not just firing the event. Please fix this or change the misleading function name.

**Comment by astei** [comment]
Doing some spring cleaning, there's been no response on this so I'm closing this PR for now. If you are still interested let me know and let's find an alternative way to implement this.

Note that I recently did #324 which might handle at least some of the use cases for `LoginFailEvent` in `DisconnectEvent` (which, arguably, is where the functionality ought to have been).

**Pull request closed without merging**